extension/app/background.ts

```diff
@@ -1,3 +1,173 @@
+
+      default:
+        console.error(`Unknown message type: ${message.type}.`);
+    }
+  }
+);
+
+/**
+ * Authenticate the user using Auth0.
+ */
+const authenticate = async (
+  sendResponse: (auth: Auth0AuthorizeResponse | AuthBackgroundResponse) => void
+) => {
+  if (!AUTH0_CLIENT_ID || !AUTH0_CLIENT_DOMAIN) {
+    console.error("Auth0 client ID or domain is missing.");
+    return;
+  }
+
+  // First we call /authorize endpoint to get the authorization code (PKCE flow).
+  const redirectUrl = chrome.identity.getRedirectURL();
+  const { codeVerifier, codeChallenge } = await generatePKCE();
+  const options = {
+    client_id: AUTH0_CLIENT_ID,
+    response_type: "code",
+    scope: "openid profile email offline_access",
+    redirect_uri: redirectUrl,
+    audience: AUTH0_AUDIENCE,
+    code_challenge_method: "S256",
+    code_challenge: codeChallenge,
+  };
+
+  const queryString = new URLSearchParams(options).toString();
+  const authUrl = `https://${AUTH0_CLIENT_DOMAIN}/authorize?${queryString}`;
+
+  chrome.identity.launchWebAuthFlow(
+    { url: authUrl, interactive: true },
+    async (redirectUrl) => {
+      if (chrome.runtime.lastError) {
+        console.error(`Auth error: ${chrome.runtime.lastError.message}`);
+        sendResponse({ success: false });
+        return;
+      }
+      if (!redirectUrl || redirectUrl.includes("error")) {
+        console.error(`Auth error in redirect URL: ${redirectUrl}`);
+        sendResponse({ success: false });
+        return;
+      }
+
+      const url = new URL(redirectUrl);
+      const queryParams = new URLSearchParams(url.search);
+      const authorizationCode = queryParams.get("code");
+
+      if (authorizationCode) {
+        // Once we have the code we call /token endpoint to exchange it for tokens.
+        const data = await exchangeCodeForTokens(
+          authorizationCode,
+          codeVerifier
+        );
+        sendResponse(data);
+      } else {
+        console.error(`No authorization code in redirect URL: ${redirectUrl}`);
+        sendResponse({ success: false });
+      }
+    }
+  );
+};
+
+/**
+ *  Exchange authorization code for tokens
+ */
+const exchangeCodeForTokens = async (
+  code: string,
+  codeVerifier: string
+): Promise<Auth0AuthorizeResponse | AuthBackgroundResponse> => {
+  try {
+    if (!AUTH0_CLIENT_ID || !AUTH0_CLIENT_DOMAIN) {
+      throw new Error("Auth0 client ID or domain is missing.");
+    }
+
+    const tokenUrl = `https://${AUTH0_CLIENT_DOMAIN}/oauth/token`;
+    const response = await fetch(tokenUrl, {
+      method: "POST",
+      headers: { "Content-Type": "application/x-www-form-urlencoded" },
+      body: new URLSearchParams({
+        grant_type: "authorization_code",
+        client_id: AUTH0_CLIENT_ID,
+        code_verifier: codeVerifier,
+        code,
+        redirect_uri: chrome.identity.getRedirectURL(),
+      }),
+    });
+
+    if (!response.ok) {
+      const data = await response.json();
```

Comment:
IIRC you don't always get JSON if it fails, you should probably do `.text()` instead to avoid crashing the background script on JSON parsing.

Reply:
Won't fail but will enter the catch and log the raw error which I think is better! 
(so not implementing unless you strongly feel it's bad)